### PR TITLE
spack external find: change default behavior

### PR DESCRIPTION
**Pull request by alalazo**
closes #29014

See https://github.com/spack/spack/issues/25353#issuecomment-1041868116

This commit changes the default behavior of
```console
$ spack external find
```
from searching all the possible packages Spack knows about to search only for the ones tagged as being a "build-tool". It also introduces a `--all` option to restore the old behavior.

Modifications:
- [x] `spack external find` looks for packages tagged as `build-tools` or `core-packages`
- [x] Tag semantic for `spack external find` is changed to be OR instead of AND
- [x] Added `--all` argument to restore the old behavior
- [x] Tagged a few packages so they are searched by default

### CODE REVIEW

**Review by sethrj — DISMISSED** [review]
Looks good, but I think there may be a few packages that need a `build-tools` tag (from looking at which packages have externals and aren't in the `spack tags` list):
- ccache
- bison
- cvs
- diffutils?
- findutils?
- gawk?
- git-annex?
- git-lfs?
- git?
- gmake
- groff
- meson
- ninja
- subversion

**Review by haampie — DISMISSED** [review]
I think this makes sense

**Comment by trws** [comment]
I mentioned this on slack, but I think it's worth mentioning here.  Limiting the list makes a *lot* of sense to me, **but** I don't think just build tools belong on the default list.  There are various "bottom" dependencies that are slow and fragile to build and can cause compatibility problems if they aren't re-used.  The main example here IMO is `openssl`, but things like `gettext` and `libicu` kinda belong here too.  If a user has cmake installed already, the difference between those being used and not will be minimal, but if they don't the difference in how likely they are to get a working build is *huge*.

**Comment by trws** [comment]
> * ccache
> * bison
> * cvs
> * diffutils?
> * findutils?
> * gawk?
> * git-annex?
> * git-lfs?
> * git?
> * gmake
> * groff
> * meson
> * ninja
> * subversion

I'd say all of these, plus:
* coreutils(!!)
* binutils
* build systems
  * scons
  * bazel
  * qmake
  * waf

Basically everything depended on by ubuntu `build-essential`:
https://packages.ubuntu.com/focal/dpkg-dev

* utilities with libs we might have to leave out for now:
  * bzip2 (utility, might not be able to do this until we can distinguish build from link dep)
  * xz utilities
  * gpg
  * perl (for build scripts, see openssl)
  * python (for build scripts, see llvm)
* patch
* tar
* fakeroot
* patchelf

**Comment by trws** [comment]
Oh, and I forgot that oldy but goodie that every gnu package needs:

* texinfo

**Comment by alalazo** [comment]
@trws Here's how it looks right now on my system:

Screenshot from 2022-02-17 19-27-07

I left a few items out of the list (e.g. `fakeroot` and `patchelf`), but we can add them later by just adding a tag to the corresponding package.

**Comment by trws** [comment]
That looks excellent!

File: var/spack/repos/builtin/packages/binutils/package.py [hunk 1]
```diff
@@ -13,6 +13,8 @@ class Binutils(AutotoolsPackage, GNUMirrorPackage):
 
     maintainers = ['alalazo']
 
+    tags = ['core-packages']
```
Comment (trws) [inline var/spack/repos/builtin/packages/binutils/package.py]: I would call this build-tools personally, this is where ld, ar and company are.  Could go either way I suppose.

Comment (alalazo) [reply var/spack/repos/builtin/packages/binutils/package.py]: Done, now it has two tags

**Review by trws — DISMISSED** [review]
This looks great, thanks @alalazo.

**Comment by trws** [comment]
Ironically a user just messaged me with one more that probably belongs in core, `openssh`.  Not as common as the rest of these, but a common run dependency for MPI for example.

File: var/spack/repos/builtin/packages/ninja/package.py [hunk 2]
```diff
@@ -14,6 +14,8 @@ class Ninja(Package):
     url      = "https://github.com/ninja-build/ninja/archive/v1.7.2.tar.gz"
     git      = "https://github.com/ninja-build/ninja.git"
 
+    tags = ['build-tools']
+
     executables = ['^ninja$']
 
     tags = ['e4s']
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/ninja/package.py]: `e4s` is going to override the build-tools tag -- and is ninja an official E4S package or just a dependency? @eugeneswalker

Comment (alalazo) [reply var/spack/repos/builtin/packages/ninja/package.py]: @sethrj Good catch, this should have been added in the list, I didn't notice the package had already a tag attribute.

**Review by sethrj — APPROVED** [review]
Yay!

**Comment by alalazo** [comment]
@scottwittenburg @zackgalbreath @tgamblin I think this is an interesting use case for pipelines. This PR just adds tags to packages - no other changes in here. It seems it's rebuilding entire stacks since it adds tags to "core" packages.

**Pull request merged**